pdns/recursordist/aggressive_nsec.cc

```diff
@@ -0,0 +1,758 @@
+{
+  WriteLock rl(d_lock);
+
+  // Force a ref until we're done removing, to avoid calling the dt
+  // while we still have the mutex locked in the else block
+  std::shared_ptr<ZoneEntry> zref;
+
+  if (subzones) {
+    d_zones.remove(zone, true);
+    updateEntriesCount();
+  }
+  else {
+    auto got = d_zones.lookup(zone);
+    if (!got || !*got || (*got)->d_zone != zone) {
+      return;
+    }
+    zref = *got;
+    std::lock_guard<std::mutex> lock((*got)->d_lock);
+    auto removed = (*got)->d_entries.size();
+    d_zones.remove(zone, false);
+    d_entriesCount -= removed;
+  }
+}
+
+void AggressiveNSECCache::prune(time_t now)
+{
+  uint64_t maxNumberOfEntries = d_maxEntries;
+  std::vector<DNSName> emptyEntries;
+
+  uint64_t erased = 0;
+  uint64_t lookedAt = 0;
+  uint64_t toLook = std::max(d_entriesCount / 5U, static_cast<uint64_t>(1U));
+
+  if (d_entriesCount > maxNumberOfEntries) {
+    uint64_t toErase = d_entriesCount - maxNumberOfEntries;
+    toLook = toErase * 5;
+    // we are full, scan at max 5 * toErase entries and stop once we have nuked enough
+
+    WriteLock rl(d_lock);
+    d_zones.visit([now, &erased, toErase, toLook, &lookedAt, &emptyEntries](const SuffixMatchTree<std::shared_ptr<ZoneEntry>>& node) {
+      if (!node.d_value || erased > toErase || lookedAt > toLook) {
+        return;
+      }
+
+      {
+        std::lock_guard<std::mutex> lock(node.d_value->d_lock);
+        auto& sidx = boost::multi_index::get<ZoneEntry::SequencedTag>(node.d_value->d_entries);
+        for (auto it = sidx.begin(); it != sidx.end(); ++lookedAt) {
+          if (erased >= toErase || lookedAt >= toLook) {
+            break;
+          }
+
+          if (it->d_ttd < now) {
+            it = sidx.erase(it);
+            ++erased;
+          }
+          else {
+            ++it;
+          }
+        }
+      }
+
+      if (node.d_value->d_entries.size() == 0) {
+        emptyEntries.push_back(node.d_value->d_zone);
+      }
+    });
+  }
+  else {
+    // we are not full, just look through 10% of the cache and nuke everything that is expired
+    WriteLock rl(d_lock);
+
+    d_zones.visit([now, &erased, toLook, &lookedAt, &emptyEntries](const SuffixMatchTree<std::shared_ptr<ZoneEntry>>& node) {
+      if (!node.d_value) {
+        return;
+      }
+
+      {
+        std::lock_guard<std::mutex> lock(node.d_value->d_lock);
+
+        auto& sidx = boost::multi_index::get<ZoneEntry::SequencedTag>(node.d_value->d_entries);
+        for (auto it = sidx.begin(); it != sidx.end(); ++lookedAt) {
+          if (lookedAt >= toLook) {
+            break;
+          }
+          if (it->d_ttd < now || lookedAt > toLook) {
+            it = sidx.erase(it);
+            ++erased;
+          }
+          else {
+            ++it;
+          }
+        }
+      }
+
+      if (node.d_value->d_entries.size() == 0) {
+        emptyEntries.push_back(node.d_value->d_zone);
+      }
+    });
+  }
+
```

Comment:
Here I would have expected a loop through `emptyEntries` to clear the empty zones.